### PR TITLE
[Bugfix][PaliGemma] Remove stale image embedding scaling

### LINKED ISSUE #52667: [Bug][PaliGemma] Image embeddings remain inverse-scaled after Gemma input scaling change
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
==============================
        System Info
==============================
OS                           : Ubuntu 22.04.5 LTS (x86_64)
GCC version                  : (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Clang version                : Could not collect
CMake version                : version 3.22.1
Libc version                 : glibc-2.35

==============================
       PyTorch Info
==============================
PyTorch version              : 2.11.0+cu130
Is debug build               : False
CUDA used to build PyTorch   : 13.0
ROCM used to build PyTorch   : N/A
XPU used to build PyTorch    : N/A

==============================
      Python Environment
==============================
Python version               : 3.12.13 (main, Jun 29 2026, 16:03:16) [GCC 11.4.0] (64-bit runtime)
Python platform              : Linux-6.12.58-82.121.amzn2023.x86_64-x86_64-with-glibc2.35

==============================
       CUDA / GPU Info
==============================
Is CUDA available            : False
CUDA runtime version         : No CUDA
CUDA_MODULE_LOADING set to   : N/A
GPU models and configuration : No CUDA
Nvidia driver version        : No CUDA
cuDNN version                : No CUDA
HIP runtime version          : N/A
MIOpen runtime version       : N/A
Is XNNPACK available         : True

==============================
          CPU Info
==============================
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  192
On-line CPU(s) list:                     0-191
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Platinum 8488C
CPU family:                              6
Model:                                   143
Thread(s) per core:                      2
Core(s) per socket:                      48
Socket(s):                               2
Stepping:                                8
BogoMIPS:                                4800.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq monitor ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx_vnni avx512_bf16 wbnoinvd ida arat avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid cldemote movdiri movdir64b md_clear serialize amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               4.5 MiB (96 instances)
L1i cache:                               3 MiB (96 instances)
L2 cache:                                192 MiB (96 instances)
L3 cache:                                210 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-47,96-143
NUMA node1 CPU(s):                       48-95,144-191
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

==============================
Versions of relevant libraries
==============================
[pip3] flashinfer-python==0.6.12
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.3.5
[pip3] nvidia-cublas==13.1.0.3
[pip3] nvidia-cuda-cccl==13.3.3.4.1
[pip3] nvidia-cuda-crt==13.3.73
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvcc==13.2.78
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cuda-tileiras==13.2.78
[pip3] nvidia-cudnn-cu13==9.19.0.56
[pip3] nvidia-cudnn-frontend==1.26.0
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-cufile==1.15.1.6
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.0
[pip3] nvidia-cutlass-dsl==4.5.2
[pip3] nvidia-cutlass-dsl-libs-base==4.5.2
[pip3] nvidia-cutlass-dsl-libs-cu13==4.5.2
[pip3] nvidia-ml-py==13.610.43
[pip3] nvidia-nccl-cu13==2.28.9
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvshmem-cu13==3.4.5
[pip3] nvidia-nvtx==13.0.85
[pip3] nvidia-nvvm==13.2.78
[pip3] pyzmq==27.1.0
[pip3] sentence-transformers==5.6.0
[pip3] tokenspeed-triton==3.8.10.post20260709
[pip3] torch==2.11.0+cu130
[pip3] torch_c_dlpack_ext==0.1.5
[pip3] torch-neuronx==2.12.3.0.23413+341d13c.dev
[pip3] torchaudio==2.11.0+cu130
[pip3] torchdata==0.11.0
[pip3] torchtitan==0.2.2
[pip3] torchvision==0.26.0+cu130
[pip3] transformers==5.13.1
[pip3] transformers-stream-generator==0.0.5
[pip3] triton==3.6.0
[conda] Could not collect

==============================
         vLLM Info
==============================
ROCM Version                 : Could not collect
vLLM Version                 : 0.24.0
vLLM Build Flags:
  CUDA Archs: Not Set; ROCm: Disabled; XPU: Disabled
GPU Topology:
  Could not collect

==============================
     Environment Variables
==============================
TORCH_NEURONX_FALLBACK_ONLY_FOR_UNIMPLEMENTED_OPS=0
TORCH_NEURONX_DISABLE_FALLBACK_EXECUTION=1
TORCH_NEURONX_MLIR_ATEN_OPS=1
TORCH_NEURONX_RETAIN_DEVICE_MODE=0
TORCH_NEURONX_ENABLE_HOST_CC=0
TORCH_NEURONX_CPP_PROCESS_GROUP=1
TORCH_NEURONX_ENABLE_STACK_TRACE=0
TORCH_NEURONX_DEBUG_DIR=/tmp/neuron_backend
TORCH_NEURONX_PRESERVE_COMPILATION_ARTIFACTS=0
TORCH_NEURONX_ENABLE_LAZY_ALLOC=1
TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_root
TORCH_NEURONX_HLO_CACHE=1
CUDA_VISIBLE_DEVICES=
CUDA_VISIBLE_DEVICES=
VLLM_USE_LAYERNAME=0
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1
```

</details>

### 🐛 Describe the bug

PR [#36787](https://github.com/vllm-project/vllm/pull/36787) moved Gemma's `sqrt(hidden_size)` normalization from `forward()` into text-only `embed_input_ids()`, but PaliGemma retained the inverse scale that previously compensated for the old common forward multiplication. Current vLLM consequently feeds projected image features into Gemma at `1 / sqrt(hidden_size)` of their intended magnitude.

Before #36787, the operations were balanced:

```text
text:  embed_tokens ------------------------> Gemma.forward * sqrt(H)
image: projector / sqrt(H) -> merge --------> Gemma.forward * sqrt(H)
```

After #36787, current vLLM scales only text:

```text
text:  embed_tokens * sqrt(H) -> merge -> Gemma.forward unchanged
image: projector / sqrt(H) ----> merge -> Gemma.forward unchanged
```

Current `main` scales text in [`gemma.py:284-308`](https://github.com/vllm-project/vllm/blob/60c3a31b12b0a735baa79684885391747771ec64/vllm/model_executor/models/gemma.py#L284-L308), but still inverse-scales images in [`paligemma.py:376-383`](https://github.com/vllm-project/vllm/blob/60c3a31b12b0a735baa79684885391747771ec64/vllm/model_executor/models/paligemma.py#L376-L383). Multimodal merge only replaces placeholder rows, and `GemmaModel.forward(inputs_embeds=...)` applies no compensating scale.

Transformers changed both halves atomically in [`0e7cb4e3`](https://github.com/huggingface/transformers/commit/0e7cb4e3ea3daaeef626148522f80b751e9559bd): Gemma text lookup gained `sqrt(H)` scaling and PaliGemma stopped dividing projected image features. Transformers 5.14.1 directly inserts raw projected image features into the already-scaled text embeddings.

For ColPali's `hidden_size=2048`, the contract reduction is:

```python
import math

hidden_size = 2048
raw_projected_image = 1.0

print(
    "before #36787:",
    raw_projected_image * hidden_size**-0.5 * math.sqrt(hidden_size),
)
print("current vLLM:", raw_projected_image * hidden_size**-0.5)
print("expected:", raw_projected_image)
```

```text
before #36787: 1.0
current vLLM: 0.022097086912079608
expected: 1.0
```

The stale factor is therefore about `0.0221`: projected image features enter Gemma at about 2.21% of their intended magnitude.

Controlled one-variable A/B against the pinned Transformers reference:

| Variant | Text cosine mean | Image cosine mean | Image p01 | Image minimum |
|---|---:|---:|---:|---:|
| Current stale inverse scale | 0.999797 | 0.903149 | 0.430221 | 0.225688 |
| Inverse scale removed only | 0.999797 | 0.987502 | 0.726556 | 0.436249 |

The request, preprocessing, checkpoint, plugin SHA, and reference were held constant; only `vision_embeddings *= hidden_size**-0.5` was removed. The remaining image mismatch is a separately measured Neuron eager-vision drift, so this one-line change should not be represented as fixing all ColPali parity issues.

Full ColPali command used for validation:

```bash
vllm serve vidore/colpali-v1.3-hf \
  --revision 7d3c8ab1c1908b32d701308fb1dfb2968d150c67 \
  --served-model-name vidore/colpali-v1.3-hf \
  --tensor-parallel-size 1 \
  --max-num-seqs 1 \
  --max-model-len 4096 \
  --trust-remote-code \
  --runner pooling \
  --convert none \
  --pooler-config.task token_embed \
  --port 20000
```

Expected behavior: PaliGemma should return raw projected image features from `embed_multimodal()`, matching Transformers' scaled-text/raw-image contract.

Suggested fix:

```diff
 def embed_multimodal(self, **kwargs: object) -> MultiModalEmbeddings:
     image_input = self._parse_and_validate_image_input(**kwargs)
     if image_input is None:
         return []
     vision_embeddings = self._process_image_input(image_input)
-    vision_embeddings = vision_embeddings * (self.config.hidden_size**-0.5)
     return vision_embeddings
```

Suggested tests: pass known projected `image_embeds` to PaliGemma and assert they are returned unchanged; add one fixed-image Transformers parity test that compares first-step logits or top-k probabilities. Current ColPali tests verify shape, L2 normalization, and weak relevance ordering, which cannot detect this pre-language-model scaling error.

The final downstream ColPali branch passes with pristine vLLM only because its host-vision path supplies raw projected features and bypasses `PaliGemma.embed_multimodal()`. The one-line patch is not part of that final plugin apply set, and the pristine pass does not validate the stale upstream method.

This regression was introduced by #36787 (`65986db6ba71abf4cf0639c5fd1477b0d8df8f5e`). Searches for PaliGemma image scaling, inverse scaling, `sqrt(hidden_size)`, `hidden_size**-0.5`, and #36787 found no matching issue or follow-up fix.

Severity: major silent correctness regression. The source contract is confirmed; broad current-main CUDA generation impact has not yet been quantified.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by ActiveSky**
## Purpose

Fixes #52667.

After Gemma moved the `sqrt(hidden_size)` scaling into text embedding lookup, PaliGemma continued to apply the old `hidden_size**-0.5` factor to projected image features. This left image embeddings at approximately `1 / sqrt(hidden_size)` of the expected magnitude. Remove the stale image-side scaling so PaliGemma passes raw projected image features to the Gemma input embedding merge, matching the Transformers contract.

This is a focused bug fix: it changes no vision projection, tokenizer, scheduler, or kernel behavior.

## Test Plan

- `uv run --no-project .venv\Scripts\python.exe -m py_compile vllm/model_executor/models/paligemma.py tests/models/multimodal/processing/test_paligemma.py`
- `VLLM_TARGET_DEVICE=cpu /root/autodl-tmp/vllm-52590/.venv/bin/python -m pytest tests/models/multimodal/processing/test_paligemma.py -v`
- Ran the same regression test against the unmodified baseline; it failed as expected because every projected embedding was still inverse-scaled.
- Ran the staged-file pre-commit checks: Ruff, format, typos, mypy, SPDX, import/configuration checks, and related hooks.
- `bash tools/pre_commit/update-dockerfile-graph.sh` passed manually. The corresponding pre-commit hook was skipped only in the Windows commit process because its absolute `/bin/bash` executable is unavailable on Windows.

## Test Result

- Targeted PaliGemma regression test: **1 passed** on the remote Linux test host with `VLLM_TARGET_DEVICE=cpu`.
- Unmodified baseline: **1 failed as expected**, confirming that the test detects the stale scaling.
- Local syntax check: **passed**.
- Relevant pre-commit hooks: **passed**.

## Duplicate-work check

Before creating this PR, I ran:

- `gh issue view 52667 --repo vllm-project/vllm --comments`
- `gh pr list --repo vllm-project/vllm --state open --search "52667 in:body"`
- `gh pr list --repo vllm-project/vllm --state open --search "PaliGemma image scaling"`

No open PR matched the issue or the affected area.

## Documentation

No documentation update is needed for this internal correctness fix.

## AI assistance

AI assistance was used to inspect the issue and call chain and to prepare the implementation and regression test. The human submitter reviewed every changed line and is responsible for the change and reported test results.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, including the issue this PR resolves.
- [x] The test plan, including commands.
- [x] The test results, including regression evidence and environment limits.
- [x] Documentation update considered; none is needed.
</details>

**BEFORE SUBMITTING, READ <https://docs.vllm.ai/en/latest/contributing>.**

### CODE REVIEW

**Review by hmellor — CHANGES_REQUESTED** [review]
~This doesn't seem right.~

~The scaling that you are deleting is specifically on the vision embeddings only, not the text embeddings. So any changes to how the text embeddings are scaled should have no effect here.~

edit: after reading the original issue this makes more sense. But the PR description is not clear at all

**Review by hmellor — APPROVED** [review]
Approving based on the description in the original issue

**Comment by ActiveSky** [comment]
Thank you very much, @hmellor, for taking the time to review this PR and for your thoughtful feedback on the PaliGemma embedding-scaling behavior. I also appreciate you revisiting the original issue and approving the PR after the clarification. Your review and support were very helpful in getting this fix merged.
